### PR TITLE
DTSPO-11610 Add private endpoint and redis version support

### PR DESCRIPTION
**Pull request by timja**
<details>
<summary>Background</summary>

The original driver for this work was the end of life of Redis 4 coming up soon.

While testing out the changes I found out that all of our Redis caches are running on a deprecated architecture called "[Cloud services (classic)](https://learn.microsoft.com/en-gb/azure/azure-cache-for-redis/cache-faq?WT.mc_id=Portal-Microsoft_Azure_Expert#what-should-i-do-with-any-instances-of-azure-cache-for-redis-that-depend-on-cloud-services--classic-)". These are deprecated for retirement. They are replaced by Virtual Machine Scale Set backed instances.

You are unable to deploy a classic Redis instance with a VMSS backed one in the same subnet.

I also noticed that there's a recommendation to move from Redis injected into the vNet to using private endpoints. One of the primary benefits is that private endpoints are supported on all SKU's.

Which means we can reduce from £120 per node to one of the smaller ones something like £17 or £41 in at least some of the environments.

</details>

Expected change for teams in their code (warning this will recreate the redis, you will likely want to do a [migration](https://learn.microsoft.com/en-gb/azure/azure-cache-for-redis/cache-vnet-migration)), by deploying a new instance, migrating existing data, and changing your app to connect to the new one. If the data doesn't matter much you could deploy a new instance and then switch users over to it without a migration.

```diff
diff --git a/infrastructure/main.tf b/infrastructure/main.tf
index a93d000..e2f2241 100644
--- a/infrastructure/main.tf
+++ b/infrastructure/main.tf
@@ -18,7 +18,12 @@ module "pcq-frontend-redis-cache" {
- source      = "git@github.com:hmcts/cnp-module-redis?ref=master"
+ source      = "git@github.com:hmcts/cnp-module-redis?ref=add-redis-version-flag" 
   product     = "${var.product}-${var.component}-redis-cache"
   location = var.location
   env      = var.env
-  subnetid = data.azurerm_subnet.core_infra_redis_subnet.id
+
+  private_endpoint_enabled = true
+  redis_version = "6"
+  business_area = "cft" # cft or sds
+  public_network_access_enabled = false
+
   common_tags  = var.common_tags
 }
```

Cache migration guide:
https://learn.microsoft.com/en-gb/azure/azure-cache-for-redis/cache-vnet-migration

Redis upgrade guide;
https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-how-to-upgrade

### CODE REVIEW

**Comment by thomast1906** [comment]
Do we want add an example folder?

**Comment by timja** [comment]
> Do we want add an example folder?

Will probably do that when I test tomorrow

File: variables.tf [hunk 1]
```diff
@@ -1,50 +1,68 @@
...
-  type    = string
-  default = "volatile-lru"
+  type        = string
+  default     = "volatile-lru"
+  description = "How Redis will select what to remove when maxmemory is reached"
 }
 
 variable "maxmemory_reserved" {
-  default = "200"
+  default     = "642"
```
Comment (timja) [inline variables.tf:44]: From Microsoft docs:
> maxmemory-reserved and maxfragmentationmemory-reserved must be between 10% (642) and 60% (3855) of the cache's total memory. Some older caches may be outside these limits. Setting a value outside of these bounds will automatically update to be within range.[Click here to learn more.](https://aka.ms/redis/memory-practices)

File: variables.tf [hunk 2]
```diff
@@ -1,50 +1,68 @@
...
+  description = "Value in megabytes reserved for non-cache usage e.g. failover"
 }
 
 variable "maxfragmentationmemory_reserved" {
-  default = "300"
+  default     = "642"
```
Comment (timja) [inline variables.tf:49]: From Microsoft docs:
> maxmemory-reserved and maxfragmentationmemory-reserved must be between 10% (642) and 60% (3855) of the cache's total memory. Some older caches may be outside these limits. Setting a value outside of these bounds will automatically update to be within range.[Click here to learn more.](https://aka.ms/redis/memory-practices)

File: variables.tf [hunk 3]
```diff
@@ -1,50 +1,68 @@
...
+  description = "Value in megabytes reserved to accommodate for memory fragmentation"
 }
 
 variable "maxmemory_delta" {
-  default = "200"
+  default     = "642"
```
Comment (timja) [inline variables.tf:54]: From Microsoft docs:
> maxmemory-reserved and maxfragmentationmemory-reserved must be between 10% (642) and 60% (3855) of the cache's total memory. Some older caches may be outside these limits. Setting a value outside of these bounds will automatically update to be within range.[Click here to learn more.](https://aka.ms/redis/memory-practices)

**Review by thomast1906 — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -50,3 +27,57 @@ resource "azurerm_key_vault_secret" "redis_access_key" {
...
+| <a name="input_private_endpoint_enabled"></a> [private\_endpoint\_enabled](#input\_private\_endpoint\_enabled) | Deploy using a private endpoint rather than vnet integration (recommended) | `bool` | `false` | no |
+| <a name="input_private_endpoint_subnet"></a> [private\_endpoint\_subnet](#input\_private\_endpoint\_subnet) | Specify your own subnet for private link integration, if you don't specify one then it will be calculated for you. | `string` | `""` | no |
+| <a name="input_product"></a> [product](#input\_product) | https://hmcts.github.io/glossary/#platform | `string` | n/a | yes |
+| <a name="input_public_network_access_enabled"></a> [public\_network\_access\_enabled](#input\_public\_network\_access\_enabled) | Whether or not public network access is allowed for this Redis Cache. `true` means this resource could be accessed by both public and private endpoint. `false` means only private endpoint access is allowed. Defaults to `true`. | `bool` | `true` | no |
+| <a name="input_redis_version"></a> [redis\_version](#input\_redis\_version) | Redis version to be deployed 4 or 6 (4 is deprecated) | `string` | `"4"` | no |
```
Comment (thomast1906) [inline README.md:71]: Do we want to remove any reference of 4?

File: variables.tf [hunk 5]
```diff
@@ -1,50 +1,99 @@
...
-  type    = list(any)
-  default = null
+  type        = list(any)
+  default     = null
+  description = "Specifies a list of Availability Zones in which this Redis Cache should be located. Changing this forces a new Redis Cache to be created."
+}
+
+variable "redis_version" {
+  type        = string
+  default     = "4"
```
Comment (thomast1906) [inline variables.tf:77]:
```suggestion
  default     = "6"
```

File: README.md (hunk 4 above)
Comment (thomast1906) [inline README.md:71]:
```suggestion
| <a name="input_redis_version"></a> [redis\_version](#input\_redis\_version) | Redis version to be deployed standard is 6 | `string` | `"4"` | no |
```

File: variables.tf (hunk 5 above)
Comment (timja) [reply variables.tf:77]: that's a breaking change and can be done in a follow up

**Comment by timja** [comment]
Your suggested changes can be done in a follow up or when all are migrated

**Pull request merged**